prow/kube/config.go

```diff
@@ -107,17 +153,46 @@ func LoadClusterConfigs(kubeconfig, buildCluster string) (configurations map[str
 			}
 		}
 		for context := range cfg.Contexts {
-			logrus.Infof("* %s", context)
+			fmt.Printf("working on context %q\n", context)
 			contextCfg, err := clientcmd.NewNonInteractiveClientConfig(*cfg, context, &clientcmd.ConfigOverrides{}, nil).ClientConfig()
 			if err != nil {
-				return nil, "", fmt.Errorf("create %s client: %v", context, err)
+				return nil, nil, fmt.Errorf("failed to create client for context %s: %v", context, err)
 			}
-			configs[context] = *contextCfg
+			logrus.WithField("context", context).Info("loaded build cluster config from cluster YAML")
+			configurations[context] = *contextCfg
 		}
+		return configurations, nil, nil
 	}
+}
 
-	if len(configs) == 0 {
-		return nil, "", errors.New("no clients found")
+func aggregateClusterConfigLoader(loaders ...clusterConfigLoader) clusterConfigLoader {
+	return func() (configurations map[string]rest.Config, defaultContext *string, err error) {
+		configurations = map[string]rest.Config{}
+		for _, loader := range loaders {
+			configurationSubset, localDefault, loadErr := loader()
+			if loadErr != nil {
+				return nil, nil, fmt.Errorf("failed to load cluster configurations: %v", loadErr)
+			}
+			for context, config := range configurationSubset {
+				if _, alreadyPresent := configurations[context]; alreadyPresent {
+					return nil, nil, fmt.Errorf("failed to load cluster configurations, context %q provided twice", context)
```

Comment:
This should not be an error, as migrating from a build-cluster file to kubecfg will require having the same entry in both locations.

The build-cluster should take precedence over kubecfg, which should take precedence over an in-cluster config.

Reply:
Why would it require that? That seems wild to me. Why would it be any harder to switch from mounting cluster yaml to kubeconfig in one deployment update? How would it help to have both at the same time with some weird overriding logic?

Reply:
Good point.

My main concern is I want to first be able to stick an image that supports both flags, and then flip from the --build-cluster to --kubeconfig flag. Rather than having to change both the flag and image at the same time.

Reply:
This behavior enables that perfectly well, you could even break it into the following steps if you really wanted to:

 - update deployment
 - continue loading build yaml
 - add mount for kubeconfig in pods
 - change flag to load kubeconfig instead of build yaml
 - remove build yaml mount